criu/cr-restore.c

```diff
@@ -1385,36 +1408,91 @@ static inline int fork_with_pid(struct pstree_item *item)
 
 	pr_info("Forking task with %d pid (flags 0x%lx)\n", pid, ca.clone_flags);
 
-	if (!(ca.clone_flags & CLONE_NEWPID)) {
-		char buf[32];
-		int len;
-		int fd = -1;
+	if (ca.item->ids)
+		pid_ns = lookup_ns_by_id(ca.item->ids->pid_ns_id, &pid_ns_desc);
 
-		if (!kdat.has_clone3_set_tid) {
-			fd = open_proc_rw(PROC_GEN, LAST_PID_PATH);
-			if (fd < 0)
-				goto err;
+	clone_flags = ca.clone_flags;
+	if (pid_ns && pid_ns->ext_key) {
+		int fd;
+
+		/* Not possible to restore into an empty PID namespace. */
+		BUG_ON(pid == INIT_PID);
+
+		/*
+		 * Restoring into an existing namespace means that CLONE_NEWPID
+		 * needs to be removed during clone() as the process will be
+		 * created in the correct PID namespace thanks to switch_ns_by_fd().
+		 */
+		clone_flags &= ~CLONE_NEWPID;
+
+		fd = inherit_fd_lookup_id(pid_ns->ext_key);
+		if (fd < 0) {
+			pr_err("Unable to find an external pidns: %s\n", pid_ns->ext_key);
+			return -1;
 		}
 
+		ret = switch_ns_by_fd(fd, &pid_ns_desc, NULL);
+		close(fd);
+
+		/*
+		 * If a process without a PID namespace is restored into
+		 * a PID namespace this tells CRIU to still handle the
+		 * process as if using CLONE_NEWPID.
+		 */
+		root_ns_mask |= CLONE_NEWPID;
+		rsti(item)->clone_flags |= CLONE_NEWPID;
+	}
+
+	if (!(clone_flags & CLONE_NEWPID)) {
+		pid_t helper_pid = -1;
+
 		lock_last_pid();
 
 		if (!kdat.has_clone3_set_tid) {
-			len = snprintf(buf, sizeof(buf), "%d", pid - 1);
-			if (write(fd, buf, len) != len) {
-				pr_perror("%d: Write %s to %s", pid, buf,
-					LAST_PID_PATH);
-				close(fd);
-				goto err_unlock;
+			if (pid_ns && pid_ns->ext_key) {
+				/*
+				 * Restoring into another namespace requires a helper
+				 * to write to LAST_PID_PATH. Using clone3() this is
+				 * so much easier and simpler. As long as CRIU supports
+				 * clone() this is needed.
+				 */
+				helper_pid = fork();
+				if (helper_pid < 0) {
+					pr_perror("Cannot fork ns_last_pid writer");
+					goto err_unlock;
+				}
+			}
+			if (helper_pid <= 0) {
+				ret = write_ns_last_pid(pid);
+				if (ret == -1 && helper_pid == -1)
+					goto err_unlock;
+				if (helper_pid == 0)
+					exit(ret);
+			}
+			if (helper_pid > 0) {
```

Comment:
I think, it would be a little easier to understand if you move `if (helper_pid <= 0)` into a helper funtion and `if (helper_pid > 0)` case into the branch with `if (pid_ns && pid_ns->ext_key)`